src/client/dfs/dfs.c

```diff
@@ -699,11 +699,39 @@ open_file(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
 		return ENOENT;
 
 fopen:
-	if (!S_ISREG(entry.mode)) {
-		if (entry.value) {
-			D_ASSERT(S_ISLNK(entry.mode));
+	if (S_ISLNK(entry.mode)) {
+		/* conditionally dereference symbolic links */
+		if (entry.value == NULL) {
+			D_ERROR("Null Symlink value\n");
+			return EIO;
+		}
+
+		if (flags & O_NOFOLLOW) {
 			D_FREE(entry.value);
+			return ELOOP;
 		}
+
+		dfs_obj_t *sym;
+
+		rc = dfs_lookup(dfs, entry.value, flags, &sym, NULL, NULL);
```

Comment:
I thought we were going to continue with the absolute paths for now, based on the root of the DFS container, until we were able to handle relative paths and the potential to go outside the container, etc.
This pre-existing code also assumes the links are absolute to the DFS container:
https://github.com/daos-stack/daos/blob/3af4e7570b183364022b5ac9fba8f5a214e452f9/src/client/dfs/dfs.c#L1973-L1982

The problem with dfs_lookup_rel is that it doesn't recursively follow links, so if you have link->link->link, then only the first will be dereferenced and returned. But dfs_lookup_rel itself should not dereference symlinks, because some other calls (specifically dfuse) expect that it will not.
But, if we are going to only allow relative paths, since open_file, open_dir, and dfs_readdir all need to be able to recursively dereference symlinks, I could create a function like this:
```
dereference_link(...)
lookup_loop:
    dfs_lookup_rel(...)
    if S_ISLNK
        goto lookup_loop
return the obj
```
And then each corresponding function can call dereference_link.
I actually implemented this solution first, so I know it's doable, but I decided that the dfs_lookup solution was simpler.

But I think the largest effort here is going to be getting relative symlinks working, which means deciding how to handle when a link goes outside the DFS container, and generally recursing backwards to handle '..'

I do think it would be best to implement the most "proper" solution rather than land something we know we're going to change later, so if you think it's best to instead only allow relative links for now, I can change this to do that instead.

Reply:
yea i know with pre-existing code we don't make that assumption. i just didn't want to land more code that does.
but we should discuss this next week with stakeholders and should have an answer what would be more appropriate then.

Reply:
Sure, thank you! 
Something else to keep in mind is that `dfuse` seems to work mostly just fine with symlinks (I'd have to re-evaluate without these changes), since the `fuse` library handles how to implement each call (e.g. stat vs lstat), and `dfuse` mostly uses `dfs_lookup_rel` and allows `fuse` to make the decision. I was wary of handling everything directly in mpifileutils, because that would be a lot more development effort for just mpifileutils, and another utilities in the future that are built on top of DFS will have to re-implement similar functionality if desired. But I could be wrong about this design decision; just something to think about.